### PR TITLE
Make take_along_axis an optional callback

### LINKED ISSUE #1366: Remove take_along_axis and take
`take` can directly call gather, no need for optional callbacks.  `take_along_axis` should have an optional callback.

### PR DESCRIPTION
**Pull request by Benjamin-Philip**
Makes take_along_axis an optional callback, using the default
implementation for BinaryBackend and EXLA, and a custom one for
Torchx.

Closes #1366.

### CODE REVIEW

**Comment by Benjamin-Philip** [comment]
I'm still yet to implement the default implementation. Hence the failing tests.

File: exla/lib/exla/defn.ex [hunk 1]
```diff
@@ -1243,43 +1243,6 @@ defmodule EXLA.Defn do
...
-    full_indices_typespec =
-      Typespec.to_shape(indices_typespec, Tuple.append(indices_shape, indices_rank))
-
-    full_indices =
-      axes_range
-      |> Enum.map(fn
-        ^axis -> Value.reshape(indices, new_axis_typespec)
-        axis -> Value.iota(state.builder, axis, new_axis_typespec)
-      end)
-      |> Value.concatenate(indices_rank, full_indices_typespec)
```
Comment (josevalim) [inline exla/lib/exla/defn.ex:1269]: FWIW, this is the main transformation you need to make in order to convert take_along_axis into a gather. :) If you do this transformation and call `gather` without the axes option, it should be equivalent to the work done here.

**Comment by Benjamin-Philip** [comment]
@josevalim, I've committed that transformation. I'm having some trouble making the optional callback work in `Defn`, which is why is failing in `Nx.Defn.GradTest` and `Nx.RandomTest`.

File: nx/lib/nx.ex [hunk 2]
```diff
@@ -14312,7 +14312,27 @@ defmodule Nx do
 
     shape = Nx.Shape.take_along_axis(tensor.shape, indices.shape, axis)
 
-    result = impl!(tensor).take_along_axis(%{tensor | shape: shape}, tensor, indices, axis)
+    out = %{tensor | shape: shape}
+
+    result =
+      Nx.Shared.optional(:take_along_axis, [tensor, indices, axis], out, fn tensor,
+                                                                            indices,
+                                                                            axis ->
```
Comment (josevalim) [inline nx/lib/nx.ex:14320]: All arguments to optional must tensors or options, so you can fix it like this:

```suggestion
      Nx.Shared.optional(:take_along_axis, [tensor, indices, [axis: axis]], out, fn
        tensor, indices, _opts ->
```

File: nx/lib/nx.ex [hunk 3]
```diff
@@ -14312,7 +14312,27 @@ defmodule Nx do
...
+          axes_range
+          |> Enum.map(fn
+            ^axis -> reshape(indices, new_axis_shape)
+            axis -> iota(new_axis_shape, axis: axis)
+          end)
+          |> concatenate(axis: rank(indices))
+
+        out = gather(tensor, full_indices)
+
+        %{out | shape: shape}
```
Comment (josevalim) [inline nx/lib/nx.ex:14334]: Interesting, why do you need to change the shape? We should probably avoid doing this!

File: nx/lib/nx/backend.ex [hunk 4]
```diff
@@ -159,6 +158,7 @@ defmodule Nx.Backend do
   @callback all_close(out :: tensor, tensor, tensor, keyword) :: tensor
   @callback top_k(out :: tensor, tensor, keyword) :: tensor
   @callback take(out :: tensor, input :: tensor, indices :: tensor, keyword) :: tensor
+  @callback take_along_axis(out :: tensor, input :: tensor, indices :: tensor, axis) :: tensor
```
Comment (josevalim) [inline nx/lib/nx/backend.ex:161]:
```suggestion
  @callback take_along_axis(out :: tensor, input :: tensor, indices :: tensor, keyword) :: tensor
```

File: nx/lib/nx.ex (hunk 3 above)
Comment (Benjamin-Philip) [reply nx/lib/nx.ex:14334]: I just copied it from the deleted impl call without thinking much about it.

**Comment by josevalim** [comment]
@Benjamin-Philip you can also remove the implementations from `expr.ex` and `grad.ex` :)

**Comment by Benjamin-Philip** [comment]
> you can also remove the implementations from expr.ex and grad.ex :)

Tried that, it still fails. I'm committing it though.

File: nx/lib/nx/backend.ex (hunk 4 above)
Comment (Benjamin-Philip) [reply nx/lib/nx/backend.ex:161]: On second thoughts, I'll revert this: since axis is the only option all the implementations directly accepted the axis.

**Comment by josevalim** [comment]
Fixed and pushed manually, thank you!

**Pull request closed without merging**